### PR TITLE
Solving crashes in async tests

### PR DESCRIPTION
**Pull request by vjpai**
Avoid thread safety issues on destructor with a proper join.

Also had been misusing EXPECT_EQ, as well as actually having an invalid
expectation on the ok field. Now it should be sane. This no longer causes a
segfault or abort trap for me.

### CODE REVIEW

File: src/cpp/server/server.cc [hunk 1]
```diff
@@ -287,12 +287,14 @@ void Server::Wait() {
 }
 
 void Server::PerformOpsOnCall(CallOpBuffer* buf, Call* call) {
-  static const size_t MAX_OPS = 8;
-  size_t nops = MAX_OPS;
-  grpc_op ops[MAX_OPS];
-  buf->FillOps(ops, &nops);
-  GPR_ASSERT(GRPC_CALL_OK ==
-             grpc_call_start_batch(call->call(), ops, nops, buf));
+  if (call->call()) {
```
Comment (ctiller) [inline src/cpp/server/server.cc]: Under what conditions can we get here?

This seems very dangerous: we'll have someone waiting on a completion that will never arrive, transforming a crash into a stuck thread.

Comment (vjpai) [inline src/cpp/server/server.cc]: We are getting here in the case reported in issue #914. The thread is shutting down, so I don't think that there's a stuckness issue. Perhaps there is a better solution to the problem.

Comment (ctiller) [inline src/cpp/server/server.cc]: To be clear - I don't believe there's a stuckness issue in the case that you're trying to fix.
I do believe that this fix opens the way for a stuckness issue in the future, as we now silently skip doing something but make an implicit promise that an event will be delivered.
Is there a point higher in the call graph that this check could be made?
Is the code validly broken and we should just issue a better error message?

Happy to sit down and look at this once we're both at the office if you'd like.

**Comment by vjpai** [comment]
Sounds good. I too worry that the mere fact that we've entered this code indicates that something horribly wrong has happened elsewhere. I missed my usual bus but am on a bus now and should be in around 10.

**Comment by ctiller** [comment]
LGTM. Will merge once the tests turn green.

**Pull request merged**